### DISCUSSION TITLE
Mock concrete types using [Mocked] attribute in AutoMoq

### DISCUSSION THREAD

**Discussion by aivascu** [AutoFixture/AutoFixture#1269] [Ideas]
As discussed #1202 sometimes it is useful to be able to tell AutoFixture to explicitly mock a concrete type.
```cs
[Theory, AutoData]
public void ConcreteTypeShouldBeAMock([Mocked] DependencyClass sut)
{
   var mock = Mock.Get(obj)
}
```
Also it can be useful to match the underlying type of a mocked instance while freezing a value.
```cs
[Theory, AutoData]
public void DependencyShouldBeSameAsMockedObject(
    [Frozen][Mock] Mock<DependencyClass> dependencyMock,
    SystemUnderTest sut)
{
    Assert.Same(dependencyMock, sut.Dependency);
}
```

**Comment by Kralizek** [comment]
> @Kralizek yes please do if you have implemented it. AFAIR when I attempted to implemented it wasn't as trivial as the implementation from above. Lets move the discussion to #1269 .

@aivascu do you have any memories regarding what stopped you?

**Reply by aivascu** [reply]
@Kralizek the implementation from the original thread suffers from the issue that the customization doesn't limit itself to a single instance resolution. Consider the following scenario:
```cs
[Theory, AutoData]
public void Foo([Mock] MyClass obj1, MyClass obj2)
{
    Moq.Get(obj1);
    Moq.Get(obj2);
}
```
In this instance the test would not fail since `obj2` would also be a mock, even though this is not explicitly defined in the test.

**Reply by Kralizek** [reply]
I guess the `MockAttribute` should try affect the resolution chain of the specified parameter only. Maybe wrapping the relay with a request specification that filters only the specific parameter?

Request specifications are a bit of a part of the system I don't know much.

**Reply by aivascu** [reply]
Yes exactly. For this to work it is necessary to create a tree of builders that affects only the parameter. Also it needs to respect the configuration of `AutoMoqCustomization` when it is applied to the `IFixture`.

Request specifications are just an implementation of the specification pattern. You can read about it [here](https://enterprisecraftsmanship.com/posts/specification-pattern-c-implementation/).

**Reply by Kralizek** [reply]
I know what a specification pattern is, I don't know how it's used in AutoFixture 😅

**Reply by aivascu** [reply]
Oh, well request specifications are used to validate whether the request satisfies the requirements of a specimen builder also to filter certain requests from entering a specimen builder.

**Comment by Kralizek** [comment]
Totally unrelated, as someone who is still learning how to properly use discussion boards in GitHub, shouldn't be this an issue? 🤔

**Reply by aivascu** [reply]
For now this is only an idea let's discuss the requirements and the API of the attribute.
